File: frontend/src/api/routes/reservations.ts
import { getReservationStatus } from '../../common/reservations'
import Api from '../Api'
import {
    AvailableHoursResponse,
    GetAvailableHoursData,
    PostReservationData,
    ReservationResponse,
    ReservationsResponse,
} from '../types'

export const getReservations = async () => {
    const resp = await Api.get<ReservationsResponse>('/reservations')
    resp.data.reservations.forEach((r) => {})
    return resp.data
}

export const getMyReservations = async () => {
    const resp = await Api.get<ReservationsResponse>('/my-reservations')
    const { reservations } = resp.data

    for (let i = 0; i < reservations.length; i++) {
        const r = reservations[i]
        const status = getReservationStatus(r)

        if (status === 'open') {
            reservations.splice(i, 1)
            reservations.unshift(r)
        }
    }

    return reservations
}

export const postReservation = async (data: PostReservationData) => {
    const resp = await Api.post<ReservationResponse>('/reservations', data)
    return resp.data
}

export const getAvailableHours = async (data: GetAvailableHoursData) => {
    const resp = await Api.post<AvailableHoursResponse>(
        '/available-hours',
        data
    )
    return resp.data
}

export const cancelReservation = async (id: number) => {
    await Api.post('/reservations/cancel', { id })
}

File: backend/src/routes/reservations.rs
use crate::auth::{RequestModOrAdmin, RequestUser};
use crate::db;
use crate::errors::{ApiError, FieldValidator};
use crate::models::table::OnlyId;

use chrono::NaiveDateTime;
use rocket_contrib::json::{Json, JsonValue};
use serde::Deserialize;

#[derive(Deserialize, Validate)]
pub struct NewReservationInfo {
    for_date: Option<NaiveDateTime>,
    people_count: Option<i32>,
}

#[post("/reservations", format = "json", data = "<new_reservation>")]
pub fn post_reservation(
    new_reservation: Json<NewReservationInfo>,
    conn: db::Conn,
    user: RequestUser,
) -> Result<JsonValue, ApiError> {
    let new_reservation = new_reservation.into_inner();

    let mut extractor = FieldValidator::validate(&new_reservation);
    let for_date = new_reservation.for_date.unwrap();
    let people_count = extractor.extract("people_count", new_reservation.people_count);

    extractor.check()?;

    db::reservations::create(&conn, for_date, people_count, &user)
        .map(|res| json!({ "reservation": res }))
}

#[post("/available-hours", format = "json", data = "<info>")]
pub fn get_available_hours(
    conn: db::Conn,
    info: Json<NewReservationInfo>,
) -> Result<JsonValue, ApiError> {
    let info = info.into_inner();

    let mut extractor = FieldValidator::validate(&info);
    let for_date = info.for_date.unwrap();
    let people_count = extractor.extract("people_count", info.people_count);

    db::reservations::get_available_hours(&conn, for_date, people_count)
        .map(|hours| json!({ "hours": hours }))
}

#[get("/reservations")]
pub fn get_reservations(
    _moderator: RequestModOrAdmin,
    conn: db::Conn,
) -> Result<JsonValue, ApiError> {
    db::reservations::get_all(&conn).map(|reservations| json!({ "reservations": reservations }))
}

#[get("/my-reservations")]
pub fn get_my_reservations(user: RequestUser, conn: db::Conn) -> Result<JsonValue, ApiError> {
    db::reservations::get_all_my(&conn, &user)
        .map(|reservations| json!({ "reservations": reservations }))
}

#[post("/reservations/cancel", format = "json", data = "<only_id>")]
pub fn cancel_reservation(
    user: RequestUser,
    only_id: Json<OnlyId>,
    conn: db::Conn,
) -> Result<(), ApiError> {
    db::reservations::cancel_reservation(&conn, only_id.id, &user)
}
